### ISSUE TITLE
False positive when trying to assign parameterized generic to type `GenericAlias`

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#4101]
`GenericAlias` is "the type of parameterized generics such as list[int]" (https://docs.python.org/3/library/types.html#types.GenericAlias).

So why is the following code fails?

```python
from types import GenericAlias

alias: GenericAlias = list[int]
```
```
error: Expression of type "Type[list[int]]" cannot be assigned to declared type "GenericAlias"
    "Type[type]" is incompatible with "Type[GenericAlias]" (reportGeneralTypeIssues)
```

**pyright**: 1.1.277

**Comment by erictraut** [comment]
Pyright is correct in generating an error here. The type of `list[int]` is not compatible with the type `GenericAlias`. 

`GenericAlias` is an internal type used at runtime. It's not meant to be used in the static type system. In the static type system, the type is `list[int]`.

**Comment by erictraut** [comment]
If your intent is to declare a type alias, you should use `typing.TypeAlias`, like this:

```python
from typing import TypeAlias

alias: TypeAlias = list[int]
```

**Comment by LeeeeT** [comment]
I want to have a type for an object with attributes `__origin__`, `__args__` and `__parameters__`. In [typeshed](https://github.com/python/typeshed/blob/8d941f0e53f5c4ff7a376c439720eb9e96aa6742/stdlib/types.pyi#L581-L586) it is declared that `GenericAlias` has all these attributes.

If I create a protocol according to this definition...
```python
class GenericAlias(Protocol):
    @property
    def __origin__(self) -> type: ...

    @property
    def __args__(self) -> tuple[Any, ...]: ...

    @property
    def __parameters__(self) -> tuple[Any, ...]: ...
```
...and try to assign `list[int]` to a variable of type `GenericAlias`, pyright says that `type[list[int]]` doesn't have these attributes, which is in fact wrong.

**Comment by erictraut** [comment]
You're conflating the static type system with the internal magic that's used to make expressions like `list[int]` work at runtime.

In the static type system, `type[list]` and `type[list[Any]]` are equivalent types. They're both instances of the `type` class, which is defined in `builtins.pyi`. The `type` class does not include `__origin__`, `__args__`, or `__parameters__`.

Not surprisingly, mypy emits the same error as pyright for the above code.

If you are using `GenericAlias`, then you're doing something very specialized (perhaps implementing a runtime type checker?). My recommendation in that case is to use a `cast` call.

**Comment by LeeeeT** [comment]
Now I understand. Thank you for the answer
